### PR TITLE
Run `XRServer::pre_render` before emitting `frame_pre_draw`

### PR DESCRIPTION
**Pull request by lyuma**
This is required to be able to use the HMD position from the `frame_pre_draw` signal in scripts.

An example usecase is making a simple portal or mirror using the Camera3D frustum mode. Currently this is not possible in XR without potentially having an old HMD position.

Needed by https://github.com/V-Sekai/avatar_vr_demo to avoid a 1 frame delay. This bug requires a mirror script and player controller to observe, which is in this project.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
This won't work @lyuma, `pre_render` in OpenXR has two main tasks:
- calling `xrWaitFrame`, which invokes OpenXRs internal timing delaying the start of rendering to bring it as close to presentation as possible
- obtaining new tracking data specific to that frame timing ensuring you have a camera position with as little lag as possible combating motion sickness

If we delay calling `pre_render` then we're calling it after we've processed our command buffers for rendering and thus we're not using the correct tracking data.

Now there is an approach spearheaded by Oculus (part of spacewarp) where you build all your command buffers for rendering, and then call `xrWaitFrame` right before submitting them, and modifying the command buffers with the new camera matrix, though that may cause a bunch of problems as we don't have a way currently to know which parts of the command buffers to change.

**Comment by clayjohn** [comment]
Bumping to 4.x as this needs to be discussed and agreed upon by the XR contributors

**Comment by addmix** [comment]
Looks like a change in 4.3 causes these changes to prevent XR from handing off properly. 
Display is never given to the headset, causing SteamVR to display the "waiting" screen indefinitely.

**Comment by lyuma** [comment]
So the underlying problem was actually improved in 4.3: the xr_server is updated at the beginning of `RendererViewport::draw_viewports`, before any Viewports have been drawn. This ought to be perfect...

However, there's still a problem with the legacy `pre_render` signal, because it is invoked in `_draw`, before `draw_viewports`, so this signal is no good. Also, generally speaking, signals have some concurrency issues and really are not ideal for hooking into the RenderingServer on another thread.

The trend in Godot is to add Callable hooks in RenderingServer for various things. In fact, there already are five hooks done as Callable.

I will have to look into these two to see if they suffice for what we need:

* [call_on_render_thread(callable: Callable)](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-call-on-render-thread)
* [void compositor_effect_set_callback(effect: RID, callback_type: CompositorEffectCallbackType, callback: Callable)](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-compositor-effect-set-callback)

These three do not help us because `call_visibility_notifiers` is run after `draw_viewports` completes. Same with `_run_post_draw_steps`.

* [void canvas_item_set_visibility_notifier(item: RID, enable: bool, area: Rect2, enter_callable: Callable, exit_callable: Callable)](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-canvas-item-set-visibility-notifier)
* [void visibility_notifier_set_callbacks(notifier: RID, enter_callable: Callable, exit_callable: Callable)](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-visibility-notifier-set-callbacks)
* [void request_frame_drawn_callback(callable: Callable)](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-request-frame-drawn-callback)

Basically, I think best would be to make a proposal that adds a signal or Callable when a viewport is about to be rendered (and again when it finishes). The other one that could be useful might be for an individual VisualInstance being rendered. That would allow a script to update uniforms, or perhaps even synchronously run a compute shader to update some properties when it's about to be drawn.

I'll try and do some research to see if a CompositorEffect can be used as a hook in lieu of a better notification that "I'm about to draw a specific viewport", or if there is another place that currently exists where we could update a uniform after xr_server but before the viewport contents are drawn.

Either way, this gets into proposal territory and isn't something that we can fix in an old PR.

**Comment by lyuma** [comment]
This PR is obsolete and no longer necessary as of the rendering viewport refactor from 4.3 I think

For future reference, it is sufficient to XRServer.add_interface a custom XRInterfaceExtension to receive a `_pre_render()` event. just make sure to return true from `is_initialized()`... and understand that it will be called from the RenderingServer thread.

```gd
class PreRenderHookInterface:
	extends XRInterfaceExtension

	signal pre_render()

	func _init():
		XRServer.add_interface(self)

	func _get_name() -> StringName:
		return &"PreRenderHookInterface"

	func _is_initialized() -> bool:
		return true

	func _pre_render() -> void:
		#print("DOING PRE RENDER!!")
		pre_render.emit()


func _ready():
	# I'm using meta to keep track of this as a singleton but feel free to use an actual singleton or autoload pattern.
	if not XRServer.has_meta(&"pre_render_hook_interface"):
		XRServer.set_meta(&"pre_render_hook_interface", PreRenderHookInterface.new())
	XRServer.get_meta(&"pre_render_hook_interface").pre_render.connect(frame_pre_draw)
```
